featurebyte/routes/user_defined_function/controller.py

```diff
@@ -0,0 +1,226 @@
+
+        Parameters
+        ----------
+        document_id: PydanticObjectId
+            UserDefinedFunction id
+        data: UserDefinedFunctionUpdate
+            UserDefinedFunction update payload
+
+        Returns
+        -------
+        UserDefinedFunctionModel
+            Updated user_defined_function object
+
+        Raises
+        ------
+        DocumentUpdateError
+            If user defined function used in any saved feature
+        """
+        # check if user defined function exists
+        document = await self.service.get_document(document_id=document_id)
+
+        # check if function used in any saved feature
+        features = await self.feature_service.list_documents(
+            query_filter={"user_defined_function_ids": {"$in": [document_id]}},
+        )
+        if features["total"]:
+            feature_names = [doc["name"] for doc in features["data"]]
+            raise DocumentUpdateError(
+                f"User defined function used by saved feature(s): {feature_names}"
+            )
+
+        func_param_mapping = OrderedDict()
+        for func_param in document.function_parameters:
+            func_param_mapping[func_param.name] = func_param
+
+        for func_param in data.function_parameters:
+            if func_param.name in func_param_mapping:
+                func_param_mapping[func_param.name] = func_param
+            else:
+                raise DocumentUpdateError(f"Function parameter not exists: {func_param.name}")
+
+        function_parameters = list(func_param_mapping.values())
+        if function_parameters == document.function_parameters:
+            raise DocumentUpdateError("No changes found in function parameters")
+
+        updated_document = await self.service.update_document(
+            document_id=document_id,
+            data=UserDefinedFunctionServiceUpdate(function_parameters=function_parameters),
+        )
+        return cast(UserDefinedFunctionModel, updated_document)
+
+    async def delete_user_defined_function(self, document_id: PydanticObjectId) -> None:
+        """
+        Delete UserDefinedFunction from persistent
+
+        Parameters
+        ----------
+        document_id: PydanticObjectId
+            UserDefinedFunction id to be deleted
+
+        Raises
+        ------
+        DocumentDeletionError
+            If user defined function used in any saved feature
+        """
+        # check if user defined function exists
+        await self.service.get_document(document_id=document_id)
+
+        # check if function used in any saved feature
+        features = await self.feature_service.list_documents(
+            query_filter={"user_defined_function_ids": {"$in": [document_id]}},
+        )
+        if features["total"]:
+            feature_names = [doc["name"] for doc in features["data"]]
+            raise DocumentDeletionError(
+                f"User defined function used by saved feature(s): {feature_names}"
+            )
```

Comment:
This parts seems to be used more than once. Might be useful to refactor as a helper function.